[8.0][base_multi_image][product_multi_image] Multi images for products split in abstract and concrete models

yajo opened the pull request:
According to https://github.com/OdooCommunityWidgets/website_multi_image/issues/34#issuecomment-157990715, I'm pushing this PR that continues job from https://github.com/OCA/product-attribute/pull/57 to discuss about design, but it is a WIP right now.

pedrobaeza committed:
[MIG] product_multi_image: Module totally redone from old product_images

pedrobaeza committed:
[FIX] product_multi_image: PEP8

yajo committed:
Split model in abstract & concrete.

yajo committed:
Some code cleanup and simplification.

yajo committed:
Update contributors.

yajo committed:
Fix some errors in code.

yajo committed:
Move field upper, so readers read the comment before.

yajo committed:
New-style reduced copyright header.

yajo committed:
Improved manifest and README for base module.

pedrobaeza commented:
The name should be _base_multi_image_ and I'm not sure if the repo target should be this or server-tools, and let _product_multi_image_ for here. Anway, this is a WIP and we can let here the discussion.

pedrobaeza commented on multi_image_base/README.rst:
```diff
@@ -0,0 +1,92 @@
+
+To develop a module based on this one:
+
+* See module ``multi_image_product`` as an example.
```
_product_multi_image_: always first the name of the "area".

elicoidal replied:
product_image_multi?

elicoidal commented on multi_image_base/README.rst:
```diff
@@ -0,0 +1,92 @@
+Multiple images base
+====================
+
+This module was written to extend the functionality of any model to support
```
This modules extends the...

elicoidal commented on multi_image_base/README.rst:
```diff
@@ -0,0 +1,92 @@
+====================
+
+This module was written to extend the functionality of any model to support
+having multiple images (a gallery) attached to it and allow you to manage them.
```
remove 'having'

elicoidal commented on multi_image_base/models/image.py:
```diff
@@ -0,0 +1,169 @@
+    image_medium = fields.Binary(
+        "Medium-sized image",
+        compute="_get_image_sizes",
+        help="Medium-sized image. It is automatically resized as a 128 x "
```
Carriage Return after 'as a' to keep the size words together

elicoidal commented on multi_image_base/models/image.py:
```diff
@@ -0,0 +1,169 @@
+        help="Medium-sized image. It is automatically resized as a 128 x "
+             "128 px image, with aspect ratio preserved, only when the image "
+             "exceeds one of those sizes. Use this field in form views or "
+             "some kanban views.")
```
remove 'some'

elicoidal commented on product_multi_image/README.rst:
```diff
@@ -0,0 +1,41 @@
+Multiple images in products
```
I miss the license badge

yajo replied:
This is now fixed.

elicoidal commented on product_multi_image/README.rst:
```diff
@@ -0,0 +1,41 @@
+* Rafael Blasco <rafabn@antiun.com>
+* Jairo Llopis <yajo.sk8@gmail.com>
+
+Maintainer
```
I miss the bug tracker section

yajo replied:
This is now fixed too.

yajo committed:
Spell checking.

yajo committed:
Refactor

Move most code to base module, move to api.multi, fix some bugs.

yajo commented:
I refactored most code to make it easier to inherit, and moving as much as possible to the base module, following the method learned with https://github.com/OCA/server-tools/pull/313. It fixed some bugs by the way, but there is one remaining that I'd thank help:
- Create a product.
- Go to images tab.
- Add one.
- That image gets set as the product's image.

Now the reverse way does not work:
- Create a product.
- Set it an image.
- No multi_image record is created, but the product gets the image!

yajo committed:
Working version.

yajo committed:
Add post_init_hook.

It takes care of moving images from the main model to the gallery
submodel, making the first image in the set always the main one. It's
implemented in
:meth:`~.multi_image_base.hooks.post_init_hook_for_submodules` for
better reusability across submodules.

yajo commented:
Finally it works (I hope). Ready to review!

rafaelbn commented:
Travis fails

yajo committed:
Add images and license key.

yajo committed:
Provide base template for website carousel galleries.

yajo committed:
Fix linter warning.

yajo committed:
Need website to add styles.

yajo committed:
Revert to v7 api to avoid conflicts.

yajo commented:
I had to write 4a43237be6c6db9cefe1430103340bb836bd32c3 to avoid conflicts caused by https://github.com/odoo/odoo/issues/10799. Let's see if it gets :green_heart: now.

rafaelbn commented:
Thanks! :green_apple: please @pedrobaeza could you change label WIP by need review!

dreispt commented:
The `multi_image_base` should be renamed as Pedro suggested, and split into it's own PR on OCA/server-tools.
Also, this is a backend feature, so I'm puzzled at the `website` dependency.

yajo commented:
The name is alright because odoo names for modules are `<section>_<functionality>`. In this case `base` indicates the functionality of the `multi_image` section, not the other way around.

Also I kept both modules in 1 PR because one does nothing without the another. I understand the base one should go to server-tools, but please review this and when I have enough :+1: I will split it before merging, so you can test now in runbot and so.

pedrobaeza commented:
@Yajo, you can say what you want, but the name is not correct. All is prefixed with `base`, not suffixed. If not, I challenge you to show me one 8.0 standard module with the `base` word at the end. Please stop debating all the reviews you receive and accept humbly the suggestions made by the reviewers.

yajo committed:
Rename to base_multi_image.

yajo commented:
Hmm you are right, I did not know that exception. About the website dependency it's about the usage of `assets_frontend` view.

dreispt commented:
The multi_image module is a generic feature, not specifically linked to products.
So I insist it should go in OCA/server-tools.

dreispt commented:
The `website` dependency is not needed then: you can add the assets to web instead of website.

yajo commented:
@dreispt 

> The multi_image module is a generic feature, not specifically linked to products.
> So I insist it should go in OCA/server-tools.

I agree. But could you (and others) please test and _conditionally approve_ this before I split? It's just because one module without the other don't make too much sense and will break bots.

> The website dependency is not needed then: you can add the assets to web instead of website.

I'm afraid the [`assets_frontend` view is defined in `website`](https://github.com/odoo/odoo/blob/8.0/addons/website/views/website_templates.xml#L14). Anyway, it's not too common to want to have an image gallery for something you don't want to show up in your website, don't you think? This is another case of perfect vs practical. I think a module that just adds that template would be an overkill.

pedrobaeza commented:
I don't agree about the website part, @Yajo. You can have multiple images because you synchronize Prestashop/Magento products, and you only want to maintain them. For that, you don't need website.

yajo commented:
Do you want 2 modules then?

dreispt commented:
@Yajo you mean 3?

yajo commented:
yes, sorry.

dreispt commented:
Can't you just replace the `website` dependency by a `web` one?

yajo commented on base_multi_image/views/assets.xml:
```diff
@@ -0,0 +1,16 @@
+<openerp>
+<data>
+
+<template id="assets_frontend" inherit_id="website.assets_frontend">
```
@dreispt Here's why I need website (or split module).

dreispt commented:
@Yajo check https://github.com/OCA/web/blob/9.0/help_popup/views/popup_help_view.xml

yajo commented on base_multi_image/views/owner_templates.xml:
```diff
@@ -0,0 +1,54 @@
+<data>
+
+<!-- Call this template in your owner's website page to get a carousel.
+     It needs a `multi_image_owner` variable set. -->
```
@dreispt These templates are to reuse in frontend website, not backend, so assets must go in frontend.

yajo replied:
See usage example in https://github.com/OCA/event/pull/33/files#diff-068a0c4a89bc9582a1c4cf77d01be469R32

yajo replied:
Since this is a base module, it provides tools to reuse in submodules, like these templates. These templates are useless without the attached CSS, and you cannot attach it without the website module. I hope you understand the dependency is **required**, so either I split the module or it stays there.

dreispt commented:
I don't understand we we need frontend features in a backend module. Confusing...
Front end features should go in their own modules and use a prefix `website_`

rafaelbn commented:
Hi @Yajo , please which is the state of this PR, please could you give a resume? Thanks

@dreispt we don't need here front-end features, **this module is done to provide base functionalities to future modules as product_multi_image, event_multi_image, ....** always in back-end. Of course when we will have such modules as event_multi_image (which depend on this one) we will do website_event_multi_image to show this multi images in front-end.

Note: like this module OCA has more base_\* modules like https://github.com/OCA/partner-contact/tree/8.0/base_contact, .... I think idea is clear.

yajo commented:
Well after confusions have been cleared, let me split this into 3 modules and PR. Thanks everyone!

yajo commented:
Thanks everyone! :wink: This is now properly split. Please review: https://github.com/OCA/server-tools/pull/374, https://github.com/OCA/website/pull/171, https://github.com/OCA/product-attribute/pull/135, https://github.com/OCA/event/pull/31, https://github.com/OCA/event/pull/33.

dreispt commented:
@Yajo Shoudn't there be a PR for OCA/product-attribute also?

yajo commented:
True, I forgot to say it: https://github.com/OCA/product-attribute/pull/135
